test/integration/pose_1_9_sdf.cc

```diff
@@ -0,0 +1,677 @@
+
+  sdf::ParamPtr poseValueParam = poseElem->GetValue();
+  ASSERT_NE(nullptr, poseValueParam);
+
+  ASSERT_TRUE(poseValueParam->Set<ignition::math::Pose3d>(
+        Pose(1, 2, 3, 0.4, 0.5, 0.6)));
+
+  sdf::ElementPtr parentElem = poseValueParam->GetParentElement();
+  ASSERT_NE(nullptr, parentElem);
+  Pose parentElemVal;
+  ASSERT_TRUE(parentElem->Get<ignition::math::Pose3d>(
+      "", parentElemVal, Pose()));
+  EXPECT_EQ(Pose(1, 2, 3, 0.4, 0.5, 0.6), parentElemVal);
+}
+
+//////////////////////////////////////////////////
+TEST(Pose1_9, ChangingParentPoseElement)
+{
+  const double pi = 3.14159265358979323846;
+  using Pose = ignition::math::Pose3d;
+
+  sdf::ElementPtr poseElem(new sdf::Element);
+  poseElem->SetName("pose");
+  poseElem->AddValue("pose", "0 0 0   0 0 0", true);
+  poseElem->AddAttribute("relative_to", "string", "", false);
+  poseElem->AddAttribute("rotation_type", "string", "rpy_radians", false);
+  ASSERT_TRUE(poseElem->Set<ignition::math::Pose3d>(
+      Pose(1, 2, 3, 0.4, 0.5, 0.6)));
+
+  sdf::ElementPtr rpyDegreesPoseElem(new sdf::Element);
+  rpyDegreesPoseElem->SetName("pose");
+  rpyDegreesPoseElem->AddValue("pose", "0 0 0   0 0 0", true);
+  rpyDegreesPoseElem->AddAttribute("relative_to", "string", "", false);
+  rpyDegreesPoseElem->AddAttribute(
+      "rotation_type", "string", "rpy_degrees", false);
+
+  sdf::ParamPtr rpyDegreesAttrib = poseElem->GetAttribute("rotation_type");
+  ASSERT_NE(nullptr, rpyDegreesAttrib);
+  ASSERT_TRUE(rpyDegreesAttrib->Set<std::string>("rpy_degrees"));
+
+  sdf::ElementPtr rpyRadiansPoseElem(new sdf::Element);
+  rpyRadiansPoseElem->SetName("pose");
+  rpyRadiansPoseElem->AddValue("pose", "0 0 0   0 0 0", true);
+  rpyRadiansPoseElem->AddAttribute("relative_to", "string", "", false);
+  rpyRadiansPoseElem->AddAttribute(
+      "rotation_type", "string", "rpy_radians", false);
+
+  sdf::ParamPtr rpyRadiansAttrib = poseElem->GetAttribute("rotation_type");
+  ASSERT_NE(nullptr, rpyRadiansAttrib);
+  ASSERT_TRUE(rpyRadiansAttrib->Set<std::string>("rpy_radians"));
+
+  // Param from original default attibute
+  sdf::ParamPtr valParam = poseElem->GetValue();
+  ASSERT_NE(nullptr, valParam);
+
+  Pose val;
+  ASSERT_TRUE(valParam->Get<ignition::math::Pose3d>(val));
+  EXPECT_EQ(Pose(1, 2, 3, 0.4, 0.5, 0.6), val);
+
+  // Set parent to Element with rotation_type attribute as rpy_degrees
+  valParam->SetParentElement(rpyDegreesPoseElem);
+  ASSERT_TRUE(valParam->Get<ignition::math::Pose3d>(val));
+  EXPECT_EQ(Pose(1, 2, 3, 0.4 * pi / 180, 0.5 * pi / 180, 0.6 * pi / 180),
+      val);
+
+  // Set parent to Element with rotation_type attribute as rpy_radians
+  valParam->SetParentElement(rpyRadiansPoseElem);
+  ASSERT_TRUE(valParam->Get<ignition::math::Pose3d>(val));
+  EXPECT_EQ(Pose(1, 2, 3, 0.4, 0.5, 0.6), val);
+
+  // Remove parent
+  valParam->SetParentElement(nullptr);
+  EXPECT_EQ(Pose(1, 2, 3, 0.4, 0.5, 0.6), val);
+}
```

Comment:
Yeah I mentioned this case in the Considerations on top, essentially:

After assigning a parent and having its values correctly parsed, the parent `Element` could have its `rotation_type` attribute changed, but that the underlying `Param` will have no way of knowing it. I considered re-evaluating the string-ified values during `Get<>` functions, but it is a `const` function, so the values should not change at that point. Unless there is a sort of reactive way to notify children when parent's attributes change, we will have to allow these issues to exist.

```cpp
param->SetParentElement(RpyElement);
param->SetFromString("1 2 3   0.4 0.5 0.6");

auto rotationType = RpyElement->GetAttribute("rotation_type");
rotationType->Set<std::string>("q_wxyz");

ignition::math::Pose3d val;
param->Get<ignition::math::Pose3d>(val); 
// val will still be (1 2 3   0.4 0.5 0.6), instead of failing or getting set to something else
```

Reply:
I considered using `SetUpdateFunc` and `Update` for the value, every time the element has any sort of modifications. What do you think?

Reply:
Nevermind the above comment, since we can always modify attributes on the fly using their pointers, there is no guarantee that the `Update` function is called.

One interesting way that might work, is that `Param` does not hold a `ParamVariant value;` at all, and every single call of `Get<T>` will make the param parse the string value in a JIT fashion, hence taking into account of any recent changes to the parent `Element` and its attributes at all times. What do you think?

Reply:
I'm not sure what the performance implications of parsing in every call to `Get<T>` would be. It would obviously be more expensive, but since we now use the DOM API more extensively, `Get<T>` might not be called all that often. However, the other consideration is error checking during parsing. If we wait only till `Get<T>` is called to parse it the string, it might be too late to notify the user of errors.

How about adding a bool argument to `Get<T>` that's determines whether to reparse or not? It can be defaulted to `false` to keep current behavior.

Reply:
So does that mean re-parsing without setting the new value to `this->dataPtr->value`, and just returning the new value? :thinking:

Reply:
As per VC, since the act of parsing only exists in developer-land, we will opt for implementing a new `Reparse` function, and this should only be called in `parser` if any attributes have changed, before a `Get<>` call on the value.

Reply:
Added the `Reparse` function and tests as described, https://github.com/ignitionrobotics/sdformat/pull/589/commits/903610466f07495938281919037c9b793ba0b809